File: macos/Sources/ClaudeUsage/UsageService.swift
import Foundation

final class UsageService {
    private static let uuidPattern = try! Regex("^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$")

    func fetch(orgId: String, cookie: String) async throws -> UsageData {
        // Validate org_id is a proper UUID to prevent URL injection
        guard orgId.wholeMatch(of: Self.uuidPattern) != nil,
              let url = URL(string: "https://claude.ai/api/organizations/\(orgId)/usage") else {
            throw UsageError.invalidResponse
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 15

        // Chrome-impersonation headers (must match the TLS fingerprint dance)
        request.setValue(cookie, forHTTPHeaderField: "Cookie")
        request.setValue("*/*", forHTTPHeaderField: "Accept")
        request.setValue("https://claude.ai/settings/usage", forHTTPHeaderField: "Referer")
        request.setValue(
            "\"Google Chrome\";v=\"147\", \"Not.A/Brand\";v=\"8\", \"Chromium\";v=\"147\"",
            forHTTPHeaderField: "Sec-Ch-Ua"
        )
        request.setValue("?0", forHTTPHeaderField: "Sec-Ch-Ua-Mobile")
        request.setValue("\"macOS\"", forHTTPHeaderField: "Sec-Ch-Ua-Platform")
        request.setValue("empty", forHTTPHeaderField: "Sec-Fetch-Dest")
        request.setValue("cors", forHTTPHeaderField: "Sec-Fetch-Mode")
        request.setValue("same-origin", forHTTPHeaderField: "Sec-Fetch-Site")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let http = response as? HTTPURLResponse else {
            throw UsageError.invalidResponse
        }

        switch http.statusCode {
        case 200:
            return try JSONDecoder().decode(UsageData.self, from: data)
        case 401, 403:
            throw UsageError.authFailed
        default:
            throw UsageError.httpError(http.statusCode)
        }
    }
}

File: windows-app/src-tauri/src/usage_service.rs
use crate::models::{UsageData, UsageError};
use once_cell::sync::Lazy;
use regex::Regex;
use reqwest::Client;
use std::time::Duration;

static UUID_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$").unwrap()
});

static CLIENT: Lazy<Client> = Lazy::new(|| {
    Client::builder()
        .timeout(Duration::from_secs(15))
        .user_agent(
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 \
             (KHTML, like Gecko) Chrome/147.0.0.0 Safari/537.36",
        )
        .build()
        .expect("failed to build reqwest client")
});

pub async fn fetch_usage(org_id: &str, cookie: &str) -> Result<UsageData, UsageError> {
    if !UUID_RE.is_match(org_id) {
        return Err(UsageError::InvalidResponse);
    }

    // Defensive: strip CR/LF to prevent header-injection if config.json was tampered with.
    let cookie = cookie.replace(['\r', '\n'], "");

    let url = format!("https://claude.ai/api/organizations/{}/usage", org_id);

    let resp = CLIENT
        .get(&url)
        .header("Cookie", &cookie)
        .header("Accept", "*/*")
        .header("Referer", "https://claude.ai/settings/usage")
        .header(
            "Sec-Ch-Ua",
            "\"Google Chrome\";v=\"147\", \"Not.A/Brand\";v=\"8\", \"Chromium\";v=\"147\"",
        )
        .header("Sec-Ch-Ua-Mobile", "?0")
        .header("Sec-Ch-Ua-Platform", "\"Windows\"")
        .header("Sec-Fetch-Dest", "empty")
        .header("Sec-Fetch-Mode", "cors")
        .header("Sec-Fetch-Site", "same-origin")
        .send()
        .await
        .map_err(|e| UsageError::Network(e.to_string()))?;

    let status = resp.status().as_u16();
    match status {
        200 => resp
            .json::<UsageData>()
            .await
            .map_err(|_| UsageError::InvalidResponse),
        401 | 403 => Err(UsageError::AuthFailed),
        other => Err(UsageError::HttpError(other)),
    }
}
